src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp

```diff
@@ -434,39 +434,52 @@ bool IpcStream::Read(void *lpBuffer, const uint32_t nBytesToRead, uint32_t &nByt
-        else
-        {
-            DWORD dwError = GetLastError();
-            if (dwError == ERROR_IO_PENDING)
+            // we are using Overlapped IO so
+            // this is expected and not an error
+            fSuccess = GetOverlappedResultEx(_hPipe,
+                                            overlap,
+                                            &nNumberOfBytesRead,
+                                            timeoutMs,
+                                            false) != 0;
+            if (!fSuccess)
             {
-                DWORD dwWait = WaitForSingleObject(_oOverlap.hEvent, (DWORD)timeoutMs);
-                if (dwWait == WAIT_OBJECT_0)
-                {
-                    // get the result
-                    fSuccess = GetOverlappedResult(_hPipe,
-                                                   overlap,
-                                                   &nNumberOfBytesRead,
-                                                   true) != 0;
-                }
-                else
+                DWORD dwError = GetLastError();
+                switch (dwError)
                 {
-                    // cancel IO and ensure the cancel happened
-                    if (CancelIo(_hPipe))
-                    {
-                        // check if the async write beat the cancellation
-                        fSuccess = GetOverlappedResult(_hPipe, overlap, &nNumberOfBytesRead, true) != 0;
-                    }
+                    case WAIT_IO_COMPLETION:
+                        // We aren't using IO Completion ports so we shouldn't see this...
+                    case ERROR_IO_INCOMPLETE:
+                        // should only happen if timeout is 0
+                        // this isn't technically an error, but the user requested a 0 timeout and the work hasn't been
+                        // completed yet, so we'll cancel.
+                    case WAIT_TIMEOUT:
+                        // We didn't complete the write in time... cancel the IO
+                        {
+                            fSuccess = CancelIoEx(_hPipe, overlap) != 0;
+                            if (!fSuccess)
+                            {
+                                DWORD dwCancelError = GetLastError();
+                            }
+                            else
+                            {
+                                fSuccess = GetOverlappedResult(_hPipe, overlap, &nNumberOfBytesRead, false) != 0;
+                                // failure either means we successfully cancelled the IO or something else went wrong.
+                                // not worth checking since we can't recover either way.
+                            }
+                        }
+                        break;
+                    default:
+                        break;
                 }
             }
         }
-        // TODO: Add error handling.
+        else
+        {
+            // TODO: Add error handling.
```

Comment:
Being here means that GetOverlappedResultEx succeeded, right? What error handling should be added? Also, in my experience checked in TODOs actually mean it never gets done. If there is work here what is preventing it from being done now?

Reply:
This `todo` has been here for a long time and simply got shuffled around with these changes, but that doesn't excuse it still being there.  I'm not sure it is needed anymore.